sdks/python/apache_beam/runners/sdf_common.py

```diff
@@ -0,0 +1,161 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""This module contains Splittable DoFn logic that's common to all runners."""
+
+import uuid
+
+import apache_beam as beam
+from apache_beam import pvalue
+from apache_beam.coders import typecoders
+from apache_beam.pipeline import AppliedPTransform
+from apache_beam.pipeline import PTransformOverride
+from apache_beam.runners.common import DoFnInvoker
+from apache_beam.runners.common import DoFnSignature
+from apache_beam.transforms.core import ParDo
+from apache_beam.transforms.ptransform import PTransform
+
+
+class SplittableParDoOverride(PTransformOverride):
+  """A transform override for ParDo transformss of SplittableDoFns.
+
+  Replaces the ParDo transform with a SplittableParDo transform that performs
+  SDF specific logic.
+  """
+
+  def get_matcher(self):
+    def _matcher(applied_ptransform):
+      assert isinstance(applied_ptransform, AppliedPTransform)
+      transform = applied_ptransform.transform
+      if isinstance(transform, ParDo):
+        signature = DoFnSignature(transform.fn)
+        return signature.is_splittable_dofn()
+
+    return _matcher
+
+  def get_replacement_transform(self, ptransform):
+    assert isinstance(ptransform, ParDo)
+    do_fn = ptransform.fn
+    signature = DoFnSignature(do_fn)
+    if signature.is_splittable_dofn():
+      return SplittableParDo(ptransform)
+    else:
+      return ptransform
+
+
+class SplittableParDo(PTransform):
+  """A transform that processes a PCollection using a Splittable DoFn."""
+
+  def __init__(self, ptransform):
+    assert isinstance(ptransform, ParDo)
+    self._ptransform = ptransform
+
+  def expand(self, pcoll):
+    sdf = self._ptransform.fn
+    signature = DoFnSignature(sdf)
+    invoker = DoFnInvoker.create_invoker(signature, process_invocation=False)
+
+    element_coder = typecoders.registry.get_coder(pcoll.element_type)
+    restriction_coder = invoker.invoke_restriction_coder()
+
+    keyed_elements = (pcoll
+                      | 'pair' >> ParDo(PairWithRestrictionFn(sdf))
+                      | 'split' >> ParDo(SplitRestrictionFn(sdf))
+                      | 'explode' >> ParDo(ExplodeWindowsFn())
```

Comment:
Ah, you are exploding windows! I think somewhere above there was code that handled multiple windows...